Implement with/2 and without/2

nalundgaard opened the pull request:
Resolves https://github.com/nalundgaard/jsn/issues/35.

See README updates in the commit for implementation notes.

nalundgaard committed:
Implement with/2 and without/2

Resolves  https://github.com/nalundgaard/jsn/issues/35.

See README updates in the commit for implementation notes.

nalundgaard commented:
@mrallen1 just a quick WIP implementation that I think does all the things for with, based on checking:

```erlang
6> Object = #{<<"foo">> => #{<<"bar">> => 1, <<"baz">> => 1, <<"fum">> => [1,2,3]}}.
#{<<"foo">> =>
      #{<<"bar">> => 1,<<"baz">> => 1,<<"fum">> => [1,2,3]}}
7> jsn:with([<<"foo.bar">>, <<"foo.baz">>, <<"foo.fum.nope">>, {<<"foo">>, <<"fum">>, last},{<<"foo">>, <<"fum">>, first}], jsn:as_proplist(Object)).
[{<<"foo">>,
  [{<<"bar">>,1},{<<"baz">>,1},{<<"fum">>,[1,3]}]}]
8> jsn:with([<<"foo.bar">>, <<"foo.baz">>, <<"foo.fum.nope">>, {<<"foo">>, <<"fum">>, 2},{<<"foo">>, <<"fum">>, first}], jsn:as_proplist(Object)).
[{<<"foo">>,
  [{<<"bar">>,1},{<<"baz">>,1},{<<"fum">>,[1,2]}]}]
9> jsn:with([<<"foo.bar">>, <<"foo.baz">>, <<"foo.fum.nope">>, {<<"foo">>, <<"fum">>, 2},{<<"foo">>, <<"fum">>, first}], Object).
#{<<"foo">> =>
      #{<<"bar">> => 1,<<"baz">> => 1,<<"fum">> => [1,2]}}
```

Before I write unit tests, I am curious if you agree with this approach. Note that it takes any JSON value, and if you ask for paths in something that doesn't have paths, you get that thing back. So like, if an intermediate element of your path is an array (or value), you just get that thing back at that layer, and it doesn't dig further or raise an exception.

I'm not sure what the right approach is. Raising exceptions sounds annoying as heck, but also it sounds tedious to take a set of deep paths and give back a shallow object with only top-level matches without telling the user, "OK, I did what you asked, but you might not like it—maybe you should validate your inputs better!"

nalundgaard commented:
... Obviously, code organization, documentation, unit tests are all TODO.

jadeallenx commented:
I think the behavior of `with/2` should mimic the behavior of `maps:with/2` which is to say if a specified (possibly deep) key doesn't exist, it's ignored in the o/p.

va17 reviewed (commented):
First take LGTM. I also think we should go down the path of mimicking `maps:with/2` and not return errors, just ignore missing paths.

nalundgaard commented:
OK, sounds like my implementation is on the right track. I added `without/2` and fixed a little bit of missing functionality, including array index handling, support for atom keys in path elements, and support for passing a pre-built path element map into the function (which will definitely be needed for use on lists of things).

nalundgaard commented on src/jsn.erl:
```diff
@@ -1027,6 +1214,52 @@ compare_maps([Key | Keys], CompareFun, A, B) ->
+%%------------------------------------------------------------------------------
+get_array_index_map(PathMap, ArrayLength) ->
+    maps:fold(fun(first, V, Acc) ->
+        Acc#{1 => V};
```
👾

va17 commented on src/jsn.erl:
```diff
@@ -200,6 +202,122 @@ select(Selection, Conditions, Elements) ->
+    PathMap1 = maps:with(maps:keys(Object), PathMap0),
+    EdgePathKeys = maps:fold(fun(K, true, Acc) -> [K | Acc];
+                                (_K, _, Acc) -> Acc
+                              end,
```
👾

nalundgaard commented:
OK, implementation and unit tests done, along with basic function documentation. Will add a README section for the 2 functions as well, along with the fun caveats about optimization (pass a pre-computed path elements map when applying repeatedly) and atom keys (only supports atom keys that are in the atom table when the path elements map is constructed, and won't work if applied to structures with atom keys that are added to the atom table later (TL;DR: don't use atom keys in JSON representation).

Curious what you think of the implementation.

nalundgaard commented:
Alright—I've written the README. I think this is ready to be reviewed as soon as you have the time, @mrallen1.

jadeallenx commented on README.md:
```diff
@@ -385,6 +399,137 @@ jsn:select({value, [<<"user">>, <<"id">>]},
+  traversing the given object and paths. That is to say, if you pass in a deeply
+  nested path, such as `foo.bar.baz.fum`, but the object only has a sub-path
+  of that path, say, `foo.bar.baz = "not an object`, the sub-path will be
+  included.
```
I don't find this example to be helpful. Do you mean to say that if the object has a path like `foo.bar.baz` and the code is searching for `foo.bar.baz.fum` that the key `foo.bar.baz` _will_ be included in the output?

nalundgaard replied:
That's exactly what I am saying.

nalundgaard replied:
Did I misunderstand the desired behavior here, BTW?

```erlang
Object = #{<<"foo">> => #{<<"bar">> => "not an object"}}.
% #{<<"foo">> => #{<<"bar">> => "not an object"}}

jsn:with(['foo.bar.baz'], Object).
% (1) don't include anything because no exact match: 
%     #{}
% or
% (2) include the partial path:
%    #{<<"foo">> => #{<<"bar">> => "not an object"}}
```

If you think we should go with (1) above, I'll fix the implementation to do that and I think we can remove both of these comments, since the behavior would be consistent (only exact match of paths trigger inclusion/exclusion). Right now, the code applies (2) behavior.

jadeallenx replied:
That right. I think paths should **exactly** match to be included. Otherwise, they are ignored. Partial matches are not **exact** and so should be ignored. To me, that's what the principle of least surprise mandates.

If you want to have like `fuzzy_with` and `fuzzy_without` then the semantics you have now would make sense. Maybe if you're basic you're going to call the functions `partial_path_match_with` and `partial_path_match_without`

jadeallenx replied:
And honestly, as long as you're allowing partial matches in a path, you might as well go full regex for key spec imo.

nalundgaard replied:
Fair enough, on reflection I totally agree with you. I've updated the code to behave accordingly (and removed these README points).

jadeallenx commented on README.md:
```diff
@@ -385,6 +399,137 @@ jsn:select({value, [<<"user">>, <<"id">>]},
+  traversing the given object and paths, only if it is encountered. That is to
+  say, if you pass in a deeply nested path, such as `foo.bar.baz.fum`, but the
+  object only has a sub-path of that path, say, `foo.bar.baz = "not an object`,
+  the sub-path will **not** be removed.
```
I don't find this comment helpful either. Please reword it.

jadeallenx commented on README.md:
```diff
@@ -385,6 +399,137 @@ jsn:select({value, [<<"user">>, <<"id">>]},
+  may not be included (or excluded) as expected. If your objects have atoms,
+  it is recommended to ensure all atoms are referenced in the system (i.e., in
+  the VM's atom table) prior to calling `jsn:path_elements_map(Paths)`. More
+  generally, however, using atom keys in your object structure is generally
```
two generallys. Eliminate one :)

jadeallenx reviewed (approved):
Great. :shipit:

rmpalomino commented on src/jsn.erl:
```diff
@@ -200,6 +202,138 @@ select(Selection, Conditions, Elements) ->
 
+-spec with(paths() | path_elements_map(), json_term()) -> json_term().
+%%------------------------------------------------------------------------------
+%% @doc
```
Empty doc

rmpalomino commented on src/jsn.erl:
```diff
@@ -200,6 +202,138 @@ select(Selection, Conditions, Elements) ->
+
+-spec without(paths() | path_elements_map(), json_term()) -> json_term().
+%%------------------------------------------------------------------------------
+%% @doc
```
Empty doc

rmpalomino commented on README.md:
```diff
@@ -385,6 +399,126 @@ jsn:select({value, [<<"user">>, <<"id">>]},
+  structure. That means that if you call this function before constructing your
+  object, and you use new atoms when doing so, those atom-keyed path elements
+  may not be included (or excluded) as expected. If your objects have atoms,
+  it is recommended to ensure all atoms are referenced in the system (i.e., in
```
Is it actually feasible to check if an atom is in the atom table? I feel like like the recommendation here is better left out.

nalundgaard replied:
Yes, `binary_to_existing_atom/2` is how we check. As for ensuring that they're present, you just have to reference them in compiled code. This is not better left out IMO.

rmpalomino reviewed (approved):
Just some pedantic comments.
